### PR TITLE
Functional test tx type 5 - Subtask #678 - Close #902

### LINKED ISSUE #902: Functional tx test type 5 - Subtask 678
Parent #678

### PR DESCRIPTION
**Pull request by diego-G**
Substask #678 
Close #902

### CODE REVIEW

File: test/functional/http/post/5.dapps.js [hunk 1]
```diff
@@ -0,0 +1,497 @@
...
+	});
+	
+	describe('confirmation', function () {
+
+		shared.confirmationPhase(goodTransactions, badTransactions);
+	});
+
+	describe('validation', function () {
+
+		it('using registered name shoudl fail', function () {
```
Comment (Isabello) [inline test/functional/http/post/5.dapps.js]: should

File: test/functional/http/post/5.dapps.js [hunk 2]
```diff
@@ -0,0 +1,497 @@
...
+			transaction = node.lisk.dapp.createDapp(account.password, null, dappDuplicateNameFail);
+
+			return sendTransactionPromise(transaction).then(function (res) {
+				node.expect(res).to.have.property('success').to.be.not.ok;
+				node.expect(res).to.have.property('message').to.equal('Application name already exists: ' + dappDuplicateNameFail.name);
+				badTransactionsEnforcement.push(transaction);
+			});
+		});
+
+		it('using registered link shoudl fail', function () {
```
Comment (Isabello) [inline test/functional/http/post/5.dapps.js]: should

File: test/functional/http/get/dapps.js [hunk 3]
```diff
@@ -1,460 +1,488 @@
...
+			})
+			.then(function (res) {
+				transaction = node.lisk.dapp.createDapp(account.password, null, dapp2);
+
+				return sendTransactionPromise(transaction);
+			})
+			.then(function (res) {
+				node.expect(res).to.have.property('success').to.be.ok;
+				node.expect(res).to.have.property('transactionId').to.equal(transaction.id);	
+				transactionsToWaitFor = [];
```
Comment (SargeKhan) [inline test/functional/http/get/dapps.js]: we add `dapp1` transaction to `transactionToWaitFor` but never call `waitForConfirmation` before reassigning `transactionToWaitFor` to empty object.

File: test/functional/http/get/dapps.js [hunk 4]
```diff
@@ -1,460 +1,488 @@
...
+				return getDappsPromise(params).then(function (res) {
+					node.expect(res).to.have.property('success');
+					node.expect(res).to.have.property('dapps').that.is.an('array').and.has.lengthOf(0);
+				});
 			});
-		});
 
-		it('using name == "Lisk Guestbook" should be ok', function (done) {
-			var name = 'Lisk Guestbook';
+			it('uusing registered dapp1 name should be ok', function () {
```
Comment (SargeKhan) [inline test/functional/http/get/dapps.js]: typo: uusing

File: test/functional/http/get/dapps.js [hunk 5]
```diff
@@ -1,460 +1,488 @@
...
-			getDapps('category:asc', function (err, res) {
-				node.expect(res.body).to.have.property('success').to.be.ok;
-				node.expect(res.body).to.have.property('dapps').that.is.an('array');
-				node.expect(res.body.dapps[0].category).to.equal(0);
-				node.expect(res.body.dapps[1].category).to.equal(1);
-				done();
+				return getDappsPromise(params).then(function (res) {
+					node.expect(res).to.have.property('success').to.be.ok;
+					node.expect(res).to.have.property('dapps').that.is.an('array');
+					node.expect(res.dapps[0].category).to.be.at.least(res.dapps[1].category);
```
Comment (SargeKhan) [inline test/functional/http/get/dapps.js]: Why are we checking category with `.to.be.at.least` assertion?

File: test/functional/http/get/dapps.js [hunk 6]
```diff
@@ -1,460 +1,488 @@
...
-		}
+			it('using "category:asc" should be ok', function () {
+				var params = [
+					'category:asc'
+				];
+				
+				return getDappsPromise(params).then(function (res) {
+					node.expect(res).to.have.property('success').to.be.ok;
+					node.expect(res).to.have.property('dapps').that.is.an('array');
+					node.expect(res.dapps[0].category).to.be.at.least(res.dapps[1].category);
```
Comment (SargeKhan) [inline test/functional/http/get/dapps.js]: Why are we checking category with .to.be.at.least assertion?

Comment (diego-G) [reply test/functional/http/get/dapps.js]: Because it can also be equal.

File: test/functional/http/get/dapps.js [hunk 7]
```diff
@@ -1,460 +1,482 @@
...
+	var transactionsToWaitFor = [];
+
+	var account = node.randomAccount();
+	var dapp1 = node.randomApplication();
+	dapp1.category = 1;
+	var dapp2 = node.randomApplication();
+	dapp2.category = 2;
+	
+	before(function () {
+		var promises = [];
```
Comment (MaciejBaj) [inline test/functional/http/get/dapps.js]: Unused

File: test/functional/http/get/dapps.js [hunk 8]
```diff
@@ -1,460 +1,482 @@
...
+var getDappsCategoriesPromise = require('../../../common/apiHelpers').getDappsCategoriesPromise;
+var waitForConfirmations = require('../../../common/apiHelpers').waitForConfirmations;
+
+describe('GET /api/dapps', function () {	
+	
+	var transactionsToWaitFor = [];
+
+	var account = node.randomAccount();
+	var dapp1 = node.randomApplication();
+	dapp1.category = 1;
```
Comment (MaciejBaj) [inline test/functional/http/get/dapps.js]: Move to the before section

File: test/functional/http/get/dapps.js [hunk 9]
```diff
@@ -1,460 +1,482 @@
...
+					node.expect(res).to.have.property('success').to.be.ok;
+					node.expect(res).to.have.property('transactionId').that.is.not.empty;
+					transactionsToWaitFor.push(res.transactionId);
+				});
+				
+				return waitForConfirmations(transactionsToWaitFor);
+			});
 	});
 
 	describe('/get?id=', function () {
```
Comment (MaciejBaj) [inline test/functional/http/get/dapps.js]:
```
describe('/get')
   describe('?')
      describe('id')
```

For consistency with the other endpoints

File: test/functional/http/get/dapps.js [hunk 10]
```diff
@@ -1,460 +1,482 @@
...
-		function getDapp (id, done) {
-			http.get('/api/dapps/get?id=' + id, done);
-		}
-
-		it('using no id should fail', function (done) {
-			getDapp('', function (err, res) {
-				node.expect(res.body).to.have.property('success').to.be.not.ok;
-				node.expect(res.body).to.have.property('error').that.is.equal('String is too short (0 chars), minimum 1: #/id');
-				done();
+		it('using no id should fail', function () {
```
Comment (MaciejBaj) [inline test/functional/http/get/dapps.js]: using undefined

File: test/functional/http/get/dapps.js [hunk 11]
```diff
@@ -1,460 +1,482 @@
...
+				return getDappsPromise(params).then(function (res) {
+					node.expect(res).to.have.property('success').to.be.not.ok;
+					node.expect(res).to.have.property('error').that.is.equal('String is too long (33 chars), maximum 32: #/name');
+				});
 			});
-		});
 
-		it('using name == "Unknown" should be ok', function (done) {
-			var name = 'Unknown';
+			it('using string == "Unknown" should be ok', function () {
```
Comment (MaciejBaj) [inline test/functional/http/get/dapps.js]: Replace == with =. Please do it for all the descriptions

File: test/functional/http/get/dapps.js [hunk 12]
```diff
@@ -1,460 +1,482 @@
...
+				node.expect(res).to.have.property('success').to.be.ok;
+				node.expect(res).to.have.property('categories').that.is.an('object');
+				for (var i in node.dappCategories) {
+					node.expect(res.categories[i]).to.equal(node.dappCategories[i]);
+				}
 			});
 		});
 	});
-});
+});
```
Comment (MaciejBaj) [inline test/functional/http/get/dapps.js]: Empty line missing

File: test/functional/http/post/5.dapps.js [hunk 13]
```diff
@@ -0,0 +1,503 @@
...
+	var goodTransactionsEnforcement = [];
+
+	var account = node.randomAccount();
+	var accountNoFunds = node.randomAccount();
+	var accountMinimalFunds = node.randomAccount();
+
+	var dappDuplicate = node.randomApplication();
+	var dappDuplicateNameSuccess = node.randomApplication();
+	var dappDuplicateNameFail = node.randomApplication();
+	dappDuplicateNameSuccess.name = dappDuplicateNameFail.name;
```
Comment (MaciejBaj) [inline test/functional/http/post/5.dapps.js]: Move to the before section

File: test/functional/http/post/5.dapps.js [hunk 14]
```diff
@@ -0,0 +1,503 @@
...
+
+		return node.Promise.all(promises)
+			.then(function (results) {
+				results.forEach(function (res) {
+					node.expect(res).to.have.property('success').to.be.ok;
+					node.expect(res).to.have.property('transactionId').that.is.not.empty;
+					transactionsToWaitFor.push(res.transactionId);
+				});
+			})
+			.then(function (res) {
```
Comment (MaciejBaj) [inline test/functional/http/post/5.dapps.js]: Unused param

File: test/functional/http/post/5.dapps.js [hunk 15]
```diff
@@ -0,0 +1,503 @@
...
+				return waitForConfirmations(transactionsToWaitFor);
+			});
+	});
+
+	describe('schema validations', function () {
+
+		shared.invalidAssets(account, 'dapp', badTransactions);
+
+		describe('category', function () {
```
Comment (MaciejBaj) [inline test/functional/http/post/5.dapps.js]: Missing test for a proper category

File: test/functional/http/post/5.dapps.js [hunk 16]
```diff
@@ -0,0 +1,503 @@
...
+
+				return sendTransactionPromise(transaction).then(function (res) {
+					node.expect(res).to.have.property('success').to.be.not.ok;
+					node.expect(res).to.have.property('message').to.match(/Value 9 is greater than maximum 8$/);
+					badTransactions.push(transaction);
+				});
+			});
+		});
+
+		describe('description', function () {
```
Comment (MaciejBaj) [inline test/functional/http/post/5.dapps.js]: Missing test for a non-string description

File: test/functional/http/post/5.dapps.js [hunk 17]
```diff
@@ -0,0 +1,503 @@
...
+
+				return sendTransactionPromise(transaction).then(function (res) {
+					node.expect(res).to.have.property('success').to.be.not.ok;
+					node.expect(res).to.have.property('message').to.match(/String is too long \(161 chars\), maximum 160$/);
+					badTransactions.push(transaction);
+				});
+			});
+		});
+
+		describe('icon', function () {
```
Comment (MaciejBaj) [inline test/functional/http/post/5.dapps.js]: Missing test for a non-string icon

File: test/functional/http/post/5.dapps.js [hunk 18]
```diff
@@ -0,0 +1,503 @@
...
+
+				return sendTransactionPromise(transaction).then(function (res) {
+					node.expect(res).to.have.property('success').to.be.not.ok;
+					node.expect(res).to.have.property('message').to.equal('Invalid application icon file type');
+					badTransactions.push(transaction);
+				});
+			});
+		});
+
+		describe('link', function () {
```
Comment (MaciejBaj) [inline test/functional/http/post/5.dapps.js]: Missing test for a non-string link
Missing test for missing link / undefined

File: test/functional/http/post/5.dapps.js [hunk 19]
```diff
@@ -0,0 +1,503 @@
...
+
+				return sendTransactionPromise(transaction).then(function (res) {
+					node.expect(res).to.have.property('success').to.be.not.ok;
+					node.expect(res).to.have.property('message').to.equal('Invalid application file type');
+					badTransactions.push(transaction);
+				});
+			});
+		});
+
+		describe('name', function () {
```
Comment (MaciejBaj) [inline test/functional/http/post/5.dapps.js]: Missing test for non-string name

File: test/functional/http/post/5.dapps.js [hunk 20]
```diff
@@ -0,0 +1,503 @@
...
+
+				return sendTransactionPromise(transaction).then(function (res) {
+					node.expect(res).to.have.property('success').to.be.not.ok;
+					node.expect(res).to.have.property('message').to.match(/String is too long \(33 chars\), maximum 32$/);
+					badTransactions.push(transaction);
+				});
+			});
+		});
+
+		describe('tags', function () {
```
Comment (MaciejBaj) [inline test/functional/http/post/5.dapps.js]: Missing test for a non-string tags

File: test/functional/http/post/5.dapps.js [hunk 21]
```diff
@@ -0,0 +1,503 @@
...
+
+				return sendTransactionPromise(transaction).then(function (res) {
+					node.expect(res).to.have.property('success').to.be.not.ok;
+					node.expect(res).to.have.property('message').to.equal('Encountered duplicate tag: ' + tag + ' in application');
+					badTransactions.push(transaction);
+				});
+			});
+		});
+
+		describe('type', function () {
```
Comment (MaciejBaj) [inline test/functional/http/post/5.dapps.js]: Missing test for a non-integer type

File: test/functional/http/post/5.dapps.js [hunk 22]
```diff
@@ -0,0 +1,503 @@
...
+				})
+				.then(function (res) {
+					node.expect(res).to.have.property('success').to.be.ok;
+					node.expect(res).to.have.property('transactionId').to.equal(transaction.id);
+					// TODO: Enable when transaction pool order is fixed
+					// goodTransactions.push(transaction);
+				});
+		});
+
+		it('two different dapps with same name should be ok and only last transaction to arrive will be confirmed', function () {
```
Comment (MaciejBaj) [inline test/functional/http/post/5.dapps.js]: `only last transaction to arrive will -> should be confirmed`, same below

File: test/functional/http/post/5.dapps.js [hunk 23]
```diff
@@ -0,0 +1,503 @@
...
+				badTransactionsEnforcement.push(transaction);
+			});
+		});	
+	});
+
+	describe('confirm validation', function () {
+
+		shared.confirmationPhase(goodTransactionsEnforcement, badTransactionsEnforcement);
+	});
+});
```
Comment (MaciejBaj) [inline test/functional/http/post/5.dapps.js]: Empty line missing

File: test/functional/http/post/5.dapps.js [hunk 24]
```diff
@@ -0,0 +1,503 @@
...
+				});
+		});
+	});
+	
+	describe('confirmation', function () {
+
+		shared.confirmationPhase(goodTransactions, badTransactions);
+	});
+
+	describe('validation', function () {
```
Comment (MaciejBaj) [inline test/functional/http/post/5.dapps.js]: Any other fields that cannot be duplicated?

File: test/functional/http/post/5.dapps.js (hunk 16 above)
Comment (diego-G) [reply test/functional/http/post/5.dapps.js]: Lisk-js does not allow an integer.

File: test/functional/http/post/5.dapps.js (hunk 17 above)
Comment (diego-G) [reply test/functional/http/post/5.dapps.js]: Same as before

File: test/functional/http/post/5.dapps.js (hunk 18 above)
Comment (diego-G) [reply test/functional/http/post/5.dapps.js]: Same as before

File: test/functional/http/post/5.dapps.js (hunk 24 above)
Comment (diego-G) [reply test/functional/http/post/5.dapps.js]: The code is checking just the two of them https://github.com/LiskHQ/lisk/blob/bd22597b26a0e36a32232f2d92737a2905fe9a31/logic/dapp.js#L163
Actually I don't get that last else.

File: test/functional/http/get/dapps.js (hunk 6 above)
Comment (SargeKhan) [reply test/functional/http/get/dapps.js]: i mean we are checking whether sorting works or not. And only checking if first two elements are in correct order. That's not right.

File: test/functional/http/get/dapps.js (hunk 5 above)
Comment (SargeKhan) [reply test/functional/http/get/dapps.js]: Same expectation for ascending order and descending order?

File: test/functional/http/get/dapps.js [hunk 25]
```diff
@@ -1,460 +1,482 @@
...
-				done();
+			it('using "category:unknown" should be ok', function () {
+				var params = [
+					'category:unknown'
+				];
+								
+				return getDappsPromise(params).then(function (res) {
+					node.expect(res).to.have.property('success').to.be.ok;
+					node.expect(res).to.have.property('dapps').that.is.an('array');
+					node.expect(res.dapps[0].category).to.be.at.least(res.dapps[1].category);
```
Comment (SargeKhan) [inline test/functional/http/get/dapps.js]: and for invalid order type as well?

File: test/functional/http/post/5.dapps.js (hunk 13 above)
Comment (SargeKhan) [reply test/functional/http/post/5.dapps.js]: and kindly explain why we need to do this.

File: test/functional/http/get/dapps.js (hunk 8 above)
Comment (diego-G) [reply test/functional/http/get/dapps.js]: I prefer to keep them out of it.

File: test/functional/http/get/dapps.js (hunk 9 above)
Comment (diego-G) [reply test/functional/http/get/dapps.js]: I prefer to have less describes when there is just one. Anyway, this function should be renamed in the refactoring.

File: test/functional/http/post/5.dapps.js (hunk 19 above)
Comment (diego-G) [reply test/functional/http/post/5.dapps.js]: Same as before

File: test/functional/http/post/5.dapps.js (hunk 20 above)
Comment (diego-G) [reply test/functional/http/post/5.dapps.js]: Same as before

File: test/functional/http/post/5.dapps.js (hunk 21 above)
Comment (diego-G) [reply test/functional/http/post/5.dapps.js]: Same as before

File: test/functional/http/post/5.dapps.js (hunk 16 above)
Comment (MaciejBaj) [reply test/functional/http/post/5.dapps.js]: Overwrite the field then after a transaction is created using lisk-js

File: test/functional/http/post/5.dapps.js (hunk 24 above)
Comment (MaciejBaj) [reply test/functional/http/post/5.dapps.js]: https://github.com/LiskHQ/lisk/blob/bd22597b26a0e36a32232f2d92737a2905fe9a31/logic/dapp.js#L151
So name, link and transactionId - add a test for to check if it is impossible to register exactly the same transaction twice

File: test/functional/http/get/dapps.js (hunk 9 above)
Comment (MaciejBaj) [reply test/functional/http/get/dapps.js]: ok

**Pull request merged**